### PR TITLE
Mark "nyxt:" URLs as secure.

### PR DESCRIPTION
**Pull request by aartaka**
# Description

This marks the `nyxt:` URLs as `:secure-p`, thus allowing outside resources, like HTTP(S) pages, to link to Nyxt-specific `nyxt:` URLs. This way, one can post example URLs and make them clickable and executable in Nyxt.

# Discussion

This is one shot at #2364, and I see two options there
- Either restricting `nyxt:` URLs locally, and unlocking their query parameters to be any `read`-able value, and maybe ever non-readable ones;
- Or opening `nyxt:` URLs up to the world, while leaving their syntax restricted.

I've picked the second way, because:
- It opens more possibilities to both `nyxt:` URLs (like embedding stylesheets and other interface elements into internal pages), and to the outside world (we can now provide links to `nyxt:` pages, like `describe-class`, on our forums, for example).
- The syntax of `nyxt:` pages doesn't have much further to move, because passing non-printable objects (like buffers) will not be easily possible however much we change the syntax. Playing with the accessibility of the links is more productive, than playing with the restricted syntax in the restricted accessibility.

But does that sound sane? @Ambrevar, @aadcg, @jmercouris? Maybe we should not even change anything in the current state of affairs? Maybe we should have a proper security audit before changing the security settings of `nyxt:` pages, given their elevated privileges? Any other opinion?

A huge security problem is that marking those URLs as secure will:
- Allow embedding JS script from the outside world into `nyxt:` pages.
- These scripts, being controlled by neither Nyxt developers, not exrension/config writers, can contain arbitrary code and change in arbitrary ways.
- Executing those in the environment of `nyxt:` pages would allow executing arbitrary Lisp code from JS.
- Executing supply chain attacks on the scripts used on `nyxt:` pages could thus lead to the full control over the browser via these compromized scripts. 
Do we even need to gaze into this Pandora box, let alone open it?

# Checklist:
Everything in this checklist is required for each PR.  Please do not approve a PR that does not have all of these items.

- [X] I have pulled from master before submitting this PR
- [X] There are no merge conflicts.
- [X] I've added the new dependencies as:
  - [X] ASDF dependencies,
  - [X] Git submodules,
    ```sh
	cd /path/to/nyxt/checkout
    git submodule add https://gitlab.common-lisp.net/nyxt/py-configparser _build/py-configparser
    ```
  - [X] and Guix dependencies.
- [X] My code follows the style guidelines for Common Lisp code. See:
  - [Norvig & Pitman's Tutorial on Good Lisp Programming Style (PDF)](https://www.cs.umd.edu/~nau/cmsc421/norvig-lisp-style.pdf)
  - [Google Common Lisp Style Guide](https://google.github.io/styleguide/lispguide.xml)
- [X] I have performed a self-review of my own code.
- [ ] My code has been reviewed by at least one peer.  (The peer review to approve a PR counts.  The reviewer must download and test the code.)
- [X] Documentation:
  - [X] All my code has docstrings and `:documentation`s written in the aforementioned style.  (It's OK to skip the docstring for really trivial parts.)
  - [X] I have updated the existing documentation to match my changes.
  - [X] I have commented my code in hard-to-understand areas.
  - [ ] I have updated the `changelog.lisp` with my changes if it's anything user-facing (new features, important bug fix, compatibility breakage).
  - [X] I have added a `migration.lisp` entry for all compatibility-breaking changes.
- [X] Compilation and tests:
  - [X] My changes generate no new warnings.
  - [X] I have added tests that prove my fix is effective or that my feature works.  (If possible.)
  - [X] New and existing unit tests pass locally with my changes.

### CODE REVIEW

**Comment by Ambrevar** [comment]
> * Allow embedding JS script from the outside world into `nyxt:` pages.

How so?

What you describe sounds a big no no.  I don't really understand how we can consider this for merging then.

Another issue, anyways, is that while `nyxt://` pages are deterministic, they technically execute code.  And while most (all?) our current `nyxt://` pages are harmless, we can't predict what the user, an extension writer or even ourselves will introduce in the future.  If a `nyxt://` page executes code depending on its arg, we open a pandora box.

I suggest we refrain from going there.  My intention with #2364 was to go the other way around, that is, add a gate that prevents any non-well controlled context from entering.

**Comment by aartaka** [comment]
> > * Allow embedding JS script from the outside world into `nyxt:` pages.
> 
> How so?

Bu adding `<script>` tags :)

> What you describe sounds a big no no. I don't really understand how we can consider this for merging then.

I mean, lots of products live with this. Our situation is more dangerous, because we expose Lisp, but we're conceptually as secure as the other projects using the CDN-distributed JS libraries.

Supply chain attacks are terrible, but they are a mere consequence of depending on someone else's code. Dependencies are a mere necessity when doing complicated products. So, if the change enables using some cool JS libraries, like D3 or Mermaid, then having a risk of supply chain attacks that is as high as in all the Web-based products is enough of a trade-off, I believe.

As an additional security measure (for when/if we merge this PR), I suggest we don't use any JS dependencies unless extremely necessary in Nyxt core internal pages. This way we can be sure that we don't ship anything compromized, while allowing one to use any dependency they like in their own pages.

> Another issue, anyways, is that while `nyxt://` pages are deterministic, they technically execute code. And while most (all?) our current `nyxt://` pages are harmless, we can't predict what the user, an extension writer or even ourselves will introduce in the future. If a `nyxt://` page executes code depending on its arg, we open a pandora box.

I'm not sure I follow. What do you mean by "executes code depending on its arg"?

> I suggest we refrain from going there. My intention with #2364 was to go the other way around, that is, add a gate that prevents any non-well controlled context from entering.

So what's your suggested way, again? Not sure I get this idea of "non-well controlled context" there...

**Comment by Ambrevar** [comment]
> Bu adding `<script>` tags :)

I don't understand.  Nyxt is responsible for generating the HTML content, so how would  third-party website produce a `nyxt://` link to injects JS?  Can you provide a more complete example?



> I suggest we don't use any JS dependencies

What does this mean?

> I'm not sure I follow. What do you mean by "executes code depending on its arg"?

Take `describe-class` for instance, it generates content that depends on its arg.  So if there is a security flaw in our code, an attacker could produce a URL with the right arg the gets `describe-class` to go through the flaw branch.


> So what's your suggested way, again? Not sure I get this idea of "non-well controlled context" there...

Still have to think about it...

I'm a bit confused with your point: pure-JS supply chain attacks in regular browsers may compromise the _browsing data_ of the user, to some limited extent.
But that's nothing compared to a compromised `nyxt://` URL which may execute arbitrary Lisp, and thus compromise the entire system of the user.

**Comment by aartaka** [comment]
> > Bu adding `<script>` tags :)
> 
> I don't understand. Nyxt is responsible for generating the HTML content, so how would third-party website produce a `nyxt://` link to injects JS? Can you provide a more complete example?

No, no, no, I mean using JS on `nyxt://` pages, not using `nyxt://` pages to _produce_ JS usable elsewhere! Which means that I suggest having `nyxt://` pages that have `<script>` tags linking to JS libraries if necessary. This was not possible until this PR.

> > I suggest we don't use any JS dependencies
> 
> What does this mean?

Again, this PR unlocks the ability to link to things beyond `nyxt://` URLs, like HTTP links to JS files, for example. This allows us to use JS libraries on our internal pages, but comes with this security tradeoff of exposing Lisp to JS. Because of that, I suggest not using any JS `<script>` tags in the core internal pages, leaving this feature to the extensions.

> > I'm not sure I follow. What do you mean by "executes code depending on its arg"?
> 
> Take `describe-class` for instance, it generates content that depends on its arg. So if there is a security flaw in our code, an attacker could produce a URL with the right arg the gets `describe-class` to go through the flaw branch.

But wait, `nyxt://` URLs only allow constant data as arguments. This ensures that the content-producing code is quite deterministic (and cacheable too). This also means that one has to sweat a lot to even bake anything `eval`-like into an internal page. And `describe-class`, even with "the most dangerous args", cannot do more than show the class documentation...

> > So what's your suggested way, again? Not sure I get this idea of "non-well controlled context" there...
> 
> Still have to think about it...
> 
> I'm a bit confused with your point: pure-JS supply chain attacks in regular browsers may compromise the _browsing data_ of the user, to some limited extent. But that's nothing compared to a compromised `nyxt://` URL which may execute arbitrary Lisp, and thus compromise the entire system of the user.

Indeed, that's quite a lot. That's why I suggest:
- Unlocking `nyxt://` URLs, marking them as `:secure-p`.
- Leaving the `nyxt://` URL parser intact to make `nyxt://` URLs secure at least from the outside.
- Never using JS `<script>`s in core internal pages.
  - And thus leaving it for extension writers and individual configs, narrowing the impact and traceability of a potential security breach.

And note that there are ways to check the script integrity that are baked into `<script>`, like `integrity` attribute, checking the SHA signature of the loaded file. Which means that we can load only what we trust, at the level of file hash.

**Comment by aartaka** [comment]
@Ambrevar, I switched `nyxt:` URLs to be `display-isolated-p` instead of being an all-in `secure-p`. What this means:
- `nyxt:` URLs _cannot_ be embedded into pages with any other scheme, which kills the iframe-embedding as the biggest danger for the `secure-p` schemes.
- `nyxt:` URLs are still universally linkable, but _are not_ universally accessible/fetchable.
- `nyxt:` URLs, even being `secure-p`, were not CORS-enabled, and neither they are now.

All in all, I believe that `display-isolated-p` is the best type for `nyxt:` URLs, because it prevents malicious use, while still expanding the abilities of the scheme for much more use-cases.

**Comment by aartaka** [comment]
No, wait, `:display-isolated-p` is too restrictive for what I described :(

**Comment by aartaka** [comment]
Okay, I found the optimal set of settings: none! Seems like not marking `nyxt:` URLs as either `:secure-p` nor `:display-isolated-p` gives us the perfect set of settings that I previously attributed to `:display-isolated-p`:
- `<iframe>` embedding doesn't work.
- Redirection (I tried JS `window.location.(href|assign|replace)`, but regular code 301 should work too!) works.
- Pages are linkable from outside!

**Comment by aadcg** [comment]
I'm not sure if I have enough background to provide much insight. I think @Ambrevar is better qualified.

**Comment by aartaka** [comment]
@Ambrevar, any feedback on this iteration (removing security settings, falling back to default scheme settings WebKit provides with properties described [in this comment](https://github.com/atlas-engineer/nyxt/pull/2586#issuecomment-1325294786))?

**Comment by aartaka** [comment]
@Ambrevar, firendly ping :) Should not take more than ten minutes to review this, starting with https://github.com/atlas-engineer/nyxt/pull/2586#issuecomment-1325294786!

**Comment by Ambrevar** [comment]
Sorry, a bit flooded these days, will come back to it.

From your comment, I'd say it sounds sane, but I'd like to study the WebKitGTK doc and...

Well,, actually could you add some tests for this?  It's critical after all and it's the perfect place to check whether third-party sites can violate access or not! :)

**Comment by aartaka** [comment]
> From your comment, I'd say it sounds sane, but I'd like to study the WebKitGTK doc and...

There's not much documentation, mostly trial and error setting one security setting or another...

> Well,, actually could you add some tests for this? It's critical after all and it's the perfect place to check whether third-party sites can violate access or not! :)

I doubt that it's sufficient, but be2518cf4ddfa8d9a51f807e4e61388e41263cd4 should test the most dangerous scenario :)

**Comment by aadcg** [comment]
What's the status here?

**Comment by aartaka** [comment]
> What's the status here?

Ready to merge on my side, but @Ambrevar may disagree, so let's wait on him!

**Comment by Ambrevar** [comment]
OK, looks good to me!

Sorry for taking so long to come back to this, no excuse really considering how trivial it is :p  

Can you add a comment for the `nyxt` internal scheme definition, explaining why all other options would not work, and why "no option" does what we want?

**Comment by aartaka** [comment]
> OK, looks good to me!
> 
> Sorry for taking so long to come back to this, no excuse really considering how trivial it is :p

No problem :)

> Can you add a comment for the `nyxt` internal scheme definition, explaining why all other options would not work, and why "no option" does what we want?

Yes, done in 55328c5289978026d1b39b67fb1a0c0552080e84. Merging then?

**Comment by Ambrevar** [comment]
Yes!

File: tests/renderer-offline/nyxt-url-security.lisp [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+  ;; Same settings as HTTPS, imitating it.
+  :secure-p t)
+
+(define-test nyxt-url-not-iframe-embeddable ()
+  (let ((ready-channel (nyxt::make-channel 1)))
+    (nyxt:start :no-config t :no-auto-config t
+                :headless t
+                :socket "/tmp/nyxt-test.socket"
+                :profile "test")
+    (hooks:once-on (nyxt:after-startup-hook *browser*) (browser)
```
Comment (aadcg) [inline tests/renderer-offline/nyxt-url-security.lisp:33]: This needs to be rewritten now that #2740 has been merges no?

File: tests/renderer-offline/nyxt-url-security.lisp [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+         (:span#text-check ""))))
+  ;; Same settings as HTTPS, imitating it.
+  :secure-p t)
+
+(define-test nyxt-url-not-iframe-embeddable ()
+  (let ((ready-channel (nyxt::make-channel 1)))
+    (nyxt:start :no-config t :no-auto-config t
+                :headless t
+                :socket "/tmp/nyxt-test.socket"
+                :profile "test")
```
Comment (aadcg) [inline tests/renderer-offline/nyxt-url-security.lisp:32]: Why not just start with `:failsafe t :headless t`?

Comment (aartaka) [reply tests/renderer-offline/nyxt-url-security.lisp:33]: Because that's how other tests do it. For consistency.

And other tests do it in the way that excludes user configuration, but doesn't exclude filesystem exisstence, just in case. For reproducibility.

File: tests/renderer-offline/nyxt-url-security.lisp (hunk 1 above)
Comment (aartaka) [reply tests/renderer-offline/nyxt-url-security.lisp]: Done.

**Comment by aartaka** [comment]
Merging now that tests are adjusted?

File: tests/renderer-offline/nyxt-url-security.lisp (hunk 2 above)
Comment (aadcg) [reply tests/renderer-offline/nyxt-url-security.lisp:33]: If that's how it's done elsewhere, then I think we should follow the example.

File: source/urls.lisp [hunk 3]
```diff
@@ -402,6 +402,17 @@ guarantee of the same result."
         (error "There's no nyxt:~a internal-page defined" symbol))))
 
 (define-internal-scheme "nyxt"
+    ;; NOTE: We use no security settings because:
+    ;; - :local-p makes nyxt:// scheme inaccessible from other schemes at all.
```
Comment (aadcg) [inline source/urls.lisp:406]: If you want to emphasize, it would sound better to say: 
"- :local-p makes nyxt:// scheme completely inaccessible from other schemes."

File: source/urls.lisp [hunk 4]
```diff
@@ -402,6 +402,17 @@ guarantee of the same result."
         (error "There's no nyxt:~a internal-page defined" symbol))))
 
 (define-internal-scheme "nyxt"
+    ;; NOTE: We use no security settings because:
```
Comment (aadcg) [inline source/urls.lisp:405]: Which security settings?

File: source/urls.lisp [hunk 5]
```diff
@@ -402,6 +402,17 @@ guarantee of the same result."
...
+    ;; - :local-p makes nyxt:// scheme inaccessible from other schemes at all.
+    ;; - :display-isolated-p does not allow displaying the scheme even inside
+    ;;   the page of the same scheme.
```
Comment (aadcg) [inline source/urls.lisp:408]: Suggestion:
- :display-isolated-p does not allow loading a nyxt:// page from a page of the same scheme.

File: source/urls.lisp [hunk 6]
```diff
@@ -402,6 +402,17 @@ guarantee of the same result."
...
+    ;; - :secure-p and :cors-enabled-p are both dangerously open for a scheme
+    ;;   that allows evaluating Lisp code.
```
Comment (aadcg) [inline source/urls.lisp:410]: Open means set to true? Maybe rephrase?

File: source/urls.lisp [hunk 7]
```diff
@@ -402,6 +402,17 @@ guarantee of the same result."
...
+    ;; That's why we've gone with no security settings:
```
Comment (aadcg) [inline source/urls.lisp:411]: The comment starts with "We use no security settings because:". The lines after this one continue to list the reasons, right?

 Do we need "That's why we've gone with no security settings:"?

File: source/urls.lisp [hunk 8]
```diff
@@ -402,6 +402,17 @@ guarantee of the same result."
...
+    ;; - <iframe> embedding and exploitation doesn't work.
```
Comment (aadcg) [inline source/urls.lisp:412]: Instead  of "doesn't work", I'd say "are impossible".

File: source/urls.lisp [hunk 9]
```diff
@@ -402,6 +402,17 @@ guarantee of the same result."
...
+    ;; - Redirection (mainly tried with window.location.(href|assign|replace), but
+    ;;   regular code 301 should work too!) works.
```
Comment (aadcg) [inline source/urls.lisp:414]: I suggest "window.location.{href,assign,replace}" so that the the parenthesis introduced in the sentence don't confuse the reader.

Instead of "regular code 301", I'd say "HTTP status code 301".

File: tests/renderer-offline/nyxt-url-security.lisp [hunk 10]
```diff
@@ -0,0 +1,41 @@
+;;;; SPDX-FileCopyrightText: Atlas Engineer LLC
```
Comment (aadcg) [inline tests/renderer-offline/nyxt-url-security.lisp:1]: Shouldn't this file be called nyxt-scheme-security?

**Comment by aadcg** [comment]
I left some minor comments.

File: tests/renderer-offline/nyxt-url-security.lisp (hunk 10 above)
Comment (aartaka) [reply tests/renderer-offline/nyxt-url-security.lisp:1]: Well, it concerns all `nyxt://`-schemed URLs, and the conventional term for those (synonymous to the `nyxt-url` function creating them) is "Nyxt URLs".

File: source/urls.lisp (hunk 5 above)
Comment (aartaka) [reply source/urls.lisp:408]: I've rephrased it, but in a slightly different and wordy way, to err on the side of caution.

**Comment by aartaka** [comment]
@aadcg, all addressed, I believe!

File: source/urls.lisp [hunk 11]
```diff
@@ -402,6 +402,18 @@ guarantee of the same result."
         (error "There's no nyxt:~a internal-page defined" symbol))))
 
 (define-internal-scheme "nyxt"
+    ;; NOTE: We use no security settings (no :local-p, :secure-p etc.) because:
+    ;; - :local-p makes nyxt:// scheme completely inaccessible from other schemes.
+    ;; - :display-isolated-p does not allow embedding the nyxt:// page even
+    ;;   inside the page of the same scheme. Self-embedding might be a use-case
```
Comment (aadcg) [inline source/urls.lisp:408]: Inside "a" page of the same scheme, not "the".

File: source/urls.lisp [hunk 12]
```diff
@@ -402,6 +402,18 @@ guarantee of the same result."
         (error "There's no nyxt:~a internal-page defined" symbol))))
 
 (define-internal-scheme "nyxt"
+    ;; NOTE: We use no security settings (no :local-p, :secure-p etc.) because:
```
Comment (aadcg) [inline source/urls.lisp:405]: It is very confusing to use "no" when referring to a predicate. 

How about "we don't set security setting such as :local-p and :secure-p to true because:"

File: source/urls.lisp [hunk 13]
```diff
@@ -402,6 +402,18 @@ guarantee of the same result."
...
+    ;; - :local-p makes nyxt:// scheme completely inaccessible from other schemes.
+    ;; - :display-isolated-p does not allow embedding the nyxt:// page even
```
Comment (aadcg) [inline source/urls.lisp:407]: "a nyxt:// page".

File: source/urls.lisp [hunk 14]
```diff
@@ -402,6 +402,18 @@ guarantee of the same result."
...
+    ;;   inside the page of the same scheme. Self-embedding might be a use-case
+    ;;   in the future, thus :display-isolated-p is not suitable.
```
Comment (aadcg) [inline source/urls.lisp:409]: "thus setting :display-isolated-p to NIL (if I understood it correctly?) is not suitable."?

File: source/urls.lisp [hunk 15]
```diff
@@ -402,6 +402,18 @@ guarantee of the same result."
...
+    ;; - :secure-p and :cors-enabled-p are both dangerously permissive for a
```
Comment (aadcg) [inline source/urls.lisp:410]: "Setting :secure-p and :cors-enable-p to T..."

File: source/urls.lisp (hunk 14 above)
Comment (aartaka) [reply source/urls.lisp:409]: T.

File: source/urls.lisp [hunk 16]
```diff
@@ -414,8 +426,7 @@ guarantee of the same result."
                 ;; FIXME: This allows `find-internal-page-buffer' to find the
                 ;; buffer and `form' to have this buffer as the buffer-var.
                 (setf (url buffer) (quri:uri url))
-                (apply (form internal-page) args)))))))
-  :local-p t)
+                (apply (form internal-page) args))))))))
```
Comment (aadcg) [inline source/urls.lisp:429]: Since the comment refers to so many predicates, wouldn't it be sensible to explicitly add them? I know we'd be setting them to default values but it would make things easier to understand.

Comment (aartaka) [reply source/urls.lisp:429]: Not really. We can change slot initforms anytime, or we can add a new renderer supporting another set of security settings. Passing them literally is neither portable nor compatible with slot initforms.

File: source/urls.lisp (hunk 11 above)
Comment (aartaka) [reply source/urls.lisp:408]: 0_o

**Comment by aadcg** [comment]
> @aadcg, all addressed, I believe!

It may seem I'm being picky but I'm really having a hard time processing that comment. I left some more comments.

**Comment by aartaka** [comment]
@aadcg, I've made yet another revision. Note that the meaning of the security settings keywords shall not be explained in detain in this comment—it's the role of WebKitGTK docs and `scheme` slots documentation to explain what those settings do and what to set them to.

My comment is merely to explain why I set none of those settings to anything. API explanation is beyond the scope.

File: source/urls.lisp [hunk 17]
```diff
@@ -402,6 +402,20 @@ guarantee of the same result."
         (error "There's no nyxt:~a internal-page defined" symbol))))
 
 (define-internal-scheme "nyxt"
+    ;; NOTE: We don't set any security settings (such as :local-p, :secure-p
+    ;; etc.)  because:
+    ;; - :local-p makes nyxt:// scheme completely inaccessible from other schemes.
+    ;; - :display-isolated-p does not allow embedding a nyxt:// page even inside
+    ;;   the page of a same scheme. Self-embedding might be a use-case in the
```
Comment (aadcg) [inline source/urls.lisp:409]: You changed the wrong "the" to "a"!

":display-isolated-p does not allow embedding a nyxt:// page even inside a page of the same scheme."

File: source/urls.lisp (hunk 15 above)
Comment (aadcg) [reply source/urls.lisp:411]: You disagree with my comment?

**Comment by aadcg** [comment]
> API explanation is beyond the scope.

Absolutely.

Feel free to merge.

File: source/urls.lisp (hunk 17 above)
Comment (aartaka) [reply source/urls.lisp:409]: Yes, that's what I've been suspicious about in https://github.com/atlas-engineer/nyxt/pull/2586#discussion_r1086496083
